### DISCUSSION TITLE
API Graph - API permissions w App registration

### DISCUSSION THREAD

**Discussion by aszykm** [AzureCommunityPL/group#132] [Pomoc] [answered]
Cześć,

Mam problem z nadawaniem uprawnień do Migrosoft Graph.
Stworzyłem skrypt, ktory za pomocą API Graph dodaje uprawnienia ale nie wiem dlaczego wpadają w "Other permissions granted for [domain] " a powinny w "Configured permissions"
 
Sugerowałem się wpisem pewnego Pana, co ciekawe u mnie to nie działa:
https://winsmarts.com/how-to-grant-admin-consent-to-an-api-programmatically-e32f4a100e9d

`$ResourceGraphURI = "https://graph.microsoft.com/"

    $graphToken = [Microsoft.Azure.Commands.Common.Authentication.AzureSession]::Instance.AuthenticationFactory.Authenticate($context.Account, $context.Environment, $context.Tenant.Id.ToString(), $null, [Microsoft.Azure.Commands.Common.Authentication.ShowDialog]::Never, $null, $ResourceGraphURI).AccessToken

    $body = @{
        clientId    = $clientId
        consentType = "AllPrincipals"
        resourceId  = $resourceId
        scope       = $scopes
        startTime   = [System.DateTime]::UtcNow.AddMinutes(-5).ToString("o")
        expiryTime  = "2225-10-19T10:37:00Z"
    }

    $apiUrl = 'https://graph.microsoft.com/v1.0/oauth2PermissionGrants'

    Write-Host $apiUrl
    Invoke-RestMethod -Uri $apiUrl -Headers @{Authorization = "Bearer $($graphToken)" }  -Method POST -Body $($body | convertto-json) -ContentType "application/json"`

Miał ktoś podobny problem lub wie jak go rozwiązać ?

**Comment by smereczynski** [comment]
@k3aczyk może pewnie pomóc :)

**Comment by mgrabarz** [comment]
Co jest w $scope i $resourceId?

**Reply by aszykm** [reply]
Na przyklad:
$scope = "Directory.ReadWrite.All"
$resourceId = "ObjectId np. MS Graph"

**Comment by aszykm** [comment, marked as answer]
Udało się rozwiązać problem za pomocą nadania uprawnień podczas tworzenia aplikacji a następnie wykonanie zapytania do API.

Co ciekawe, samo zapytanie do API nie nadaje uprawnień a jedynie "udziela zgody administratoa" (z ang. grant admin consent). 
Jest to zaskakujące, ponieważ według dokumentacji, jest to nadawanie uprawnień a nie tylko udzielenie zgody:
https://docs.microsoft.com/en-us/graph/api/oauth2permissiongrant-post?view=graph-rest-1.0&tabs=http

Jeżeli kogoś interesuje to tak powinno to być wykonane:

```
$scope = "User.Read"

$msGraph = Get-AzureADServicePrincipal -All $true | ? { $_.DisplayName -match "Microsoft Graph" }

$reqGraph =  New-Object -TypeName "Microsoft.Open.AzureAD.Model.RequiredResourceAccess
$reqGraph.ResourceAppId = $msGraph.AppId
$idUserReadPerm = ($msGraph.Oauth2Permissions | ? { $_.Value -match "^"+$scope+"$"}).Id
$appPermission = New-Object -TypeName "Microsoft.Open.AzureAD.Model.ResourceAccess" -ArgumentList $idUserReadPerm,"Scope"
$reqGraph.ResourceAccess = $appPermission

New-AzureADApplication  (...)  -RequiredResourceAccess $reqGraph
```

Następnie wykonujemy zapytanie do API, które przedstawiłem przy tworzeniu dyskusji.
